### PR TITLE
Use a normalized file name for s3 package uploads

### PR DESCRIPTION
**Pull request by shaunmarx**

### CODE REVIEW

File: source/Calamari.Aws/Deployment/Conventions/UploadAwsS3Convention.cs [hunk 1]
```diff
@@ -246,9 +246,16 @@ public Task<S3UploadResult> UploadUsingPackage(Func<AmazonS3Client> clientFactor
...
+            var id = deployment.Variables.Get(SpecialVariables.Packages.PackageId(null));
+            var version = deployment.Variables.Get(SpecialVariables.Packages.PackageVersion(null));
+            var extension = Path.GetExtension(deployment.Variables.Get(SpecialVariables.Packages.OriginalPath(null)));
+            var directory = Path.GetDirectoryName(deployment.PackageFilePath);
+            var filename = $"{id}.{version}{extension}";
+            var filePath = directory != null ? Path.Combine(directory,  filename) : filename;
+            
             return CreateRequest(deployment.PackageFilePath,
-                    GetBucketKey(Path.GetDirectoryName(deployment.PackageFilePath), deployment.PackageFilePath, options), options)
+                    GetBucketKey(directory, filePath, options), options)
```
Comment (flin-8) [inline source/Calamari.Aws/Deployment/Conventions/UploadAwsS3Convention.cs]: Potential null-ref inside `GetBucketKey` if directory is `null`

File: source/Calamari.Aws/Deployment/Conventions/UploadAwsS3Convention.cs [hunk 2]
```diff
@@ -246,9 +246,16 @@ public Task<S3UploadResult> UploadUsingPackage(Func<AmazonS3Client> clientFactor
             Guard.NotNull(deployment, "Deployment may not be null");
             Guard.NotNull(options, "Package options may not be null");
             Guard.NotNull(clientFactory, "Client factory must not be null");
-
+            
+            var id = deployment.Variables.Get(SpecialVariables.Packages.PackageId(null));
```
Comment (flin-8) [inline source/Calamari.Aws/Deployment/Conventions/UploadAwsS3Convention.cs]: These lines should be extracted to a function

File: source/Calamari.Aws/Deployment/Conventions/UploadAwsS3Convention.cs [hunk 3]
```diff
@@ -246,9 +246,16 @@ public Task<S3UploadResult> UploadUsingPackage(Func<AmazonS3Client> clientFactor
...
+            var version = deployment.Variables.Get(SpecialVariables.Packages.PackageVersion(null));
```
Comment (flin-8) [inline source/Calamari.Aws/Deployment/Conventions/UploadAwsS3Convention.cs]: We really call `SpecialVariables.Packages.PackageVersion()` with `null` ?!? Is that the convention?

Comment (shaunmarx) [reply source/Calamari.Aws/Deployment/Conventions/UploadAwsS3Convention.cs]: Don't get me started :p

File: source/Calamari.Tests/AWS/S3/BucketKeyProviderFixture.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+namespace Calamari.Tests.AWS.S3
+{
+    [TestFixture]
+    public class BucketKeyProviderFixture
+    {
+        [Test]
+        public void PackageFilenameBehaviorUsesDefaultWithPrefix()
+        {
+            var provider = new BucketKeyProvider();
```
Comment (flin-8) [inline source/Calamari.Tests/AWS/S3/BucketKeyProviderFixture.cs]: Use the `SetUp()` method to new up the `sut` (system under test)

**Pull request merged**